romanesco/tasks/docker.py

```diff
@@ -0,0 +1,101 @@
+    """
+    command = ('docker', 'pull', image)
+    p = subprocess.Popen(args=command, stdout=subprocess.PIPE,
+                         stderr=subprocess.PIPE)
+    stdout, stderr = p.communicate()
+
+    if p.returncode != 0:
+        print('Error pulling docker image %s:' % image)
+        print('STDOUT: ' + stdout)
+        print('STDERR: ' + stderr)
+
+        raise Exception('Docker pull returned code {}.'.format(p.returncode))
+
+
+def _transformPath(inputs, taskInputs, inputId, tmpDir):
+    """
+    If the input specified by inputId is a filepath target, we transform it to
+    its absolute path within the docker container (underneath /data).
+    """
+    for ti in taskInputs.itervalues():
+        tiId = ti['id'] if 'id' in ti else ti['name']
+        if tiId == inputId:
+            if ti.get('target') == 'filepath':
+                rel = os.path.relpath(inputs[inputId]['script_data'], tmpDir)
+                return os.path.join('/data', rel)
+            else:
+                return inputs[inputId]['script_data']
+
+    raise Exception('No task input found with id = ' + inputId)
+
+
+def _expandArgs(args, inputs, taskInputs, tmpDir):
+    """
+    Expands arguments to the container execution if they reference input
+    data. For example, if an input has id=foo, then a container arg of the form
+    $input{foo} would be expanded to the runtime value of that input. If that
+    input is a filepath target, the file path will be transformed into the
+    location that it will be available inside the running container.
+    """
+    newArgs = []
+    regex = re.compile(r'\$input\{([^}]+)\}')
+
+    for arg in args:
+        match = re.search(regex, arg)
+        if match is None:
+            newArgs.append(arg)
+        else:
+            inputId = match.groups()[0]
+            if inputId in inputs:
+                newArgs.append(_transformPath(inputs, taskInputs, inputId,
+                                              tmpDir))
+            else:
+                newArgs.append(arg)
+
+    return newArgs
+
+def run(task, inputs, outputs, task_inputs, task_outputs, **kwargs):
+    image = task['docker_image']
+    print('Pulling docker image: ' + image)
+    _pullImage(image)
+
+    tmpDir = kwargs.get('_tmp_dir')
+    args = _expandArgs(task['container_args'], inputs, task_inputs, tmpDir)
+
+    print('Running container with args: ' + ' '.join(args))
+
+    command = ['docker', 'run']
+
+    if tmpDir:
+        command += ['-v', tmpDir + ':/data']
+
+    command += [image] + args
+    volumeMap = '%s:%s' % (os.path.abspath(tmpDir), '/data')
+
+    p = subprocess.Popen(args=command, stdout=subprocess.PIPE,
+                          stderr=subprocess.PIPE)
+    stdout, stderr = p.communicate()
+
+    outputs['_stdout'] = {
+        'format': 'string',
+        'script_data': stdout
+    }
+
+    outputs['_stderr'] = {
+        'format': 'string',
+        'script_data': stderr
+    }
```

Comment:
No stdin is not a priority feature - just brainstorming.

I suppose since we now have name and id as separate fields in outputs, we could rely on the id to match e.g. `"_stdout"` to map directly to the output stream - so I think that's a good approach. I wouldn't want `_stdout` to appear in the user interface, but it will be hidden due to the separate UI-friendly name field.

It is a bit tricky how to interpret stdout. I suppose if an output is mapped to stdout then we just capture and stream stderr. If nothing is mapped to stdout we could stream both stdout and stderr to the client. Does that seem reasonable? I agree that it is convenient (and will be common practice) to just send the true algorithm output to stdout when using docker containers, so we certainly want to support that.